src/include/sof/measure.h

```diff
@@ -0,0 +1,199 @@
+ *
+ * The context is used to store current statistics and measurement status,
+ * therefore it should be kept and reused with the following calls.
+ *
+ * \param[in] steps_log2 History length for the computation of running average,
+ *                       in the form of a power of 2: log2(steps)
+ * \return Measurement context
+ */
+struct measure_context measure_init(uint32_t steps_log2);
+
+/**
+ * \brief Invalidates and frees context
+ * \param[in] context Measurement context
+ */
+void measure_free(struct measure_context *context);
+
+/**
+ * \brief Sets the measurement reference
+ *
+ * Timing mesurements will be taken in relation to the point of calling
+ * this function.
+ *
+ * \param[in] context Measurement context
+ */
+void measure_begin(struct measure_context *context);
+
+/**
+ * \brief Takes the measurement
+ *
+ * Performs timing measurements in relation to previously set reference
+ * (see measure_begin()) and recalculates statistics.
+ *
+ * \param[in] context Measurement context
+ */
+void measure_end(struct measure_context *context);
+
+/**
+ * \brief Returns microseconds of a given statistic type
+ * \param[in] context Measurement context
+ * \param[in] stat_type Statistic type (min, avg, max)
+ * \return Time in microseconds
+ */
+uint64_t measure_get_microseconds(struct measure_context *context,
+				  int stat_type);
+
+/**
+ * \brief Returns ccount of a given statistic type
+ * \param[in] context Measurement context
+ * \param[in] stat_type Statistic type (min, avg, max)
+ * \return Number of cycles
+ */
+uint64_t measure_get_ccount(struct measure_context *context, int stat_type);
+
+/**
+ * \brief Reports the measurement
+ *
+ * Will report the measurements every interval milliseconds
+ * in the form of a trace containing average timing:
+ *
+ * "<prefix> - avg perf: %u us, %u cycles"
+ * The prefix may contain up to 2 32-bit integer parameters.
+ *
+ * \param[in] context Measurement context
+ * \param[in] interval Trace interval for the reports in milliseconds
+ * \param[in] trace_class Trace class
+ * \param[in] str_prefix_fmt String prefix format to use in the reported trace
+ * \param[in] ... Parameters for the provided prefix format (max. 2 params)
+ */
+#define measure_report(context, interval, trace_class, str_prefix_fmt, ...) \
+	do { \
+		typeof(context) _ctx = (context); \
+		uint64_t time = platform_timer_get(platform_timer); \
+		\
+		if (time - _ctx->last_report >= \
+		    clock_ms_to_ticks(PLATFORM_DEFAULT_CLOCK, (interval))) { \
+			uint32_t perf_us = measure_get_microseconds( \
+				_ctx, MEASURE_STAT_AVG); \
+			\
+			uint32_t perf_ccount = measure_get_ccount( \
+				_ctx, MEASURE_STAT_AVG); \
+			\
+			_ctx->last_report = time; \
+			trace_event((trace_class), str_prefix_fmt \
+				    PERF_REPORT_FMT, ##__VA_ARGS__, \
+				    perf_us, perf_ccount); \
+		} \
+	} while (0)
+
+#else
+
+struct measure_context {};
+static inline struct measure_context measure_init(uint32_t steps)
```

Comment:
If we take measure_context struct definition outside of #if CONFIG_PERF_MEASURE there is no need for the dummy definition here.

Reply:
Yes, provided that the context is stored as a pointer. Otherwise it will end up wasting space the way it is now.

Reply:
@slawblauciak exactly, let's store it as a pointer.

Reply:
Wrt the space used, steps and avg_shift are compile time consts. Do you really need to keep a different run-time value in every instance of the structure? Otherwise they can be removed and array of samples could be also defined as a cont. memory w/o another allocation & redirection.